torchinductor/codegen/wrapper.py

```diff
@@ -55,6 +57,7 @@ def __init__(self):
         self._names_iter = count()
         self.header = IndentedBuffer()
         self.prefix = IndentedBuffer()
+        self.allocation = DeferredIndentedBuffer()
```

Comment:
Will this move all the allocations to the top of the function?  This might not be good because it could increase peak memory usage in long functions.

Currently, we allocate right before first use, then `del` on last use (there is also a reuse algorithm).

Reply:
The FreedBuffer class is also sort of doing deferred codegen here:
https://github.com/pytorch/torchdynamo/blob/ad29d6899caca0bad4114443e244b3444c5873de/torchinductor/codegen/wrapper.py#L43


We change the `del buf0` into a `buf5 = buf0` late in the process if we want to reuse the memory for buf0.